api/clients/v2/validator/retrieval_worker.go

```diff
@@ -0,0 +1,709 @@
+		if w.getStatusCount(verifying, verified) > w.targetVerifiedCount {
+			// We've requested enough verifications
+			break
+		}
+
+		next, _ := w.downloadedChunksQueue.Dequeue()
+		reply := next.(*validatorDownloadCompleted).Reply
+		operatorID := next.(*validatorDownloadCompleted).OperatorID
+
+		w.updateChunkStatus(operatorID, verifying)
+
+		w.computePool.Submit(func() {
+			w.deserializeAndVerifyChunks(w.ctx, operatorID, reply)
+		})
+	}
+}
+
+// decodeBlob decodes the blob from the chunks.
+func (w *retrievalWorker) decodeChunks() ([]byte, error) {
+	if w.detailedLogging {
+		w.logger.Info("decoding blob", "blobKey", w.blobKey.Hex())
+	}
+
+	chunks := make([]*encoding.Frame, 0)
+	indices := make([]uint, 0)
+
+	for !w.verifiedChunksQueue.Empty() {
+		next, _ := w.verifiedChunksQueue.Dequeue()
+		operatorID := next.(*verificationCompleted).OperatorID
+		operatorChunks := next.(*verificationCompleted).chunks
+
+		assignment := w.assignments[operatorID]
+		assignmentIndices := make([]uint, len(assignment.GetIndices()))
+
+		for i, index := range assignment.GetIndices() {
+			assignmentIndices[i] = uint(index)
+		}
+
+		chunks = append(chunks, operatorChunks...)
+		indices = append(indices, assignmentIndices...)
+	}
+
+	// TODO: faster to truncate the number of chunks if we have more than we need?
+
+	w.computePool.Submit(func() {
+		w.decodeBlob(chunks, indices)
+	})
+
+	select {
+	case <-w.ctx.Done():
+		return nil, fmt.Errorf("retrieval worker context cancelled: %w", w.ctx.Err())
+	case decodeResponse := <-w.decodeResponseChan:
+		if decodeResponse.Err == nil {
+			return decodeResponse.Blob, nil
+		} else {
+			return nil, fmt.Errorf("failed to decode blob: %w", decodeResponse.Err)
+		}
+	}
+}
+
+// downloadChunks downloads the chunk data from the specified operator.
+func (w *retrievalWorker) downloadChunks(operatorID core.OperatorID) {
+	if w.detailedLogging {
+		w.logger.Debug("downloading chunks",
+			"operator", operatorID.Hex(),
+			"blobKey", w.blobKey.Hex())
+	}
+
+	// Report back to the control loop when the download started. This may be later than when
+	// the download was scheduled if there is contention for the connection pool.
+	w.downloadStartedChan <- &downloadStarted{
+		operatorID:    operatorID,
+		downloadStart: w.timeSource(),
+	}
+
+	// Unless this has been overridden for testing, this is just a call to standardDownloadChunks().
+	reply, err := w.downloadChunksFunction(w.ctx, w.blobKey, operatorID)
+
+	w.downloadCompletedChan <- &validatorDownloadCompleted{
+		OperatorID: operatorID,
+		Reply:      reply,
+		Err:        err,
+	}
+}
+
+// downloadChunksFunction is the default function used to download chunks from a validator node. This may not
+// be called if
+func (w *retrievalWorker) standardDownloadChunks(
+	_ context.Context,
+	_ v2.BlobKey,
+	operatorID core.OperatorID,
+) (*grpcnode.GetChunksReply, error) {
+
+	ctx, cancel := context.WithTimeout(w.downloadAndVerifyContext, w.config.DownloadTimeout)
```

Comment:
Would it make sense to set this timeout above, in `downloadChunks`, before calling the `downloadChunksFunction`? Then pass in the `ctx` with correctly configured timeout as a parameter, and use it directly